cs_api_io/cs_api_io.ml

```diff
@@ -1,79 +1,53 @@
 
 let send_request_exn ~verify {Api.Request.url; header; method_; data} =
-  let open Lwt.Infix in
-  let method_str =
+  Curl.global_init Curl.CURLINIT_GLOBALALL;
+  let curl = Curl.init () in
+  let (response, response_callback) = response_accumulator_factory () in
+  Curl.set_url curl url;
+  Curl.set_sslverifypeer curl verify;
+  Curl.set_writefunction curl response_callback;
+  set_headers curl header;
+  let _ =
     match method_ with
-    | Get -> "GET"
-    | Post -> "POST"
+    | Get -> Curl.set_httpget curl true
+    | Post -> (
+      match data with
+      | Raw str -> Curl.set_postfields curl str
+      | Multipart parts -> set_multipart curl parts )
   in
-  Printf.printf "Request: %s %s\n%!" method_str url;
-  let headers = Cohttp.Header.add_list (Cohttp.Header.init ()) header in
-  let url = Uri.of_string url in
-  get_ctx ~verify >>= fun ctx ->
-  let open Lwt_result.Infix in
-  match method_ with
-  | Get -> Lwt_result.ok (Cohttp_lwt_unix.Client.get ~ctx ~headers url)
-  | Post ->
-    ( match data with
-    | Raw str -> handle_raw str
-    | Multipart parts -> handle_multipart parts )
-    >>= fun (extra_headers, body) ->
-    let headers = Cohttp.Header.add_list headers extra_headers in
-    let body = Cohttp_lwt.Body.of_stream body in
-    Lwt_result.ok (Cohttp_lwt_unix.Client.post ~ctx ~body ~headers url)
+  try
+    Curl.perform curl;
+    Lwt_result.return {Response.code = Curl.get_responsecode curl; response}
+  with Curl.CurlException (_, _, s) ->
+    Lwt_result.fail ("HTTP exception : " ^ s)
 
-let send_request ?(verify = true) request =
-  Lwt.catch
-    (fun () -> send_request_exn ~verify request)
-    (function
-      | error ->
-        let message =
-          match error with
-          | Failure _ ->
-            "URL could not be reached"
-            (* Matching a specific failure results in warnings *)
-          | Tls_lwt.Tls_failure _ -> "Could not establish HTTPS connection"
-          | Unix.Unix_error (Unix.ECONNREFUSED, "connect", "") ->
-            "Could not establish network connection"
-          | Ssl.Connection_error _ -> Printexc.to_string error
-          | _ ->
-            Printf.sprintf "Unexpected exception: %s" (Printexc.to_string error)
-        in
-        Lwt_result.fail message)
+let send_request ?(verify = true) request = send_request_exn ~verify request
 
-let get_response (response, body) =
-  let open Lwt.Infix in
-  let status = Cohttp.Response.status response in
-  let code = Cohttp.Code.code_of_status status in
-  Cohttp_lwt.Body.to_string body >>= fun body_str ->
-  if Cohttp.Code.is_error code then
-    let message =
-      Printf.sprintf "%s\n%s" (Cohttp.Code.string_of_status status) body_str
-    in
-    Lwt_result.fail message
+let get_response {Response.code; response} =
+  if code < 300 then
+    Lwt_result.return !response
   else
-    Lwt_result.return body_str
+    Lwt_result.fail !response
```

Comment:
What about keeping the same sort of message as before when there's a failure, with status code and body?